forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabRepository.java

```diff
@@ -625,49 +625,63 @@ public List<CommitComment> recentCommitComments(ReadOnlyRepository localRepo, Se
     }
 
     /**
-     * Lazy fetching and caching of the commitTitleToCommits map. The first time
-     * this is called, the full map is built from the local repository. After that
-     * it's just refreshed from the server.
+     * Lazy fetching and caching of the commitTitleToCommits map. The map is only
+     * cached for one set of branches. If called with a different set, the old
+     * cache is invalidated and a new is rebuilt from scratch. In practice, an
+     * instance of this class is only ever called with the same set of branches.
+     * If that was to change we will need to rethink this.
+     * <p>
+     * The full map is built from the local repository. For incremental updates,
+     * it's refreshed until it contains the same number of commits as the local
+     * repository. The topo-order should guarantee that this gives us the same
+     * set of commits.
      */
     private final Map<String, SequencedSet<Hash>> commitTitleToCommits = new HashMap<>();
-    private boolean commitTitleToCommitsInitialized = false;
-    private ZonedDateTime lastCommitTime = ZonedDateTime.ofInstant(Instant.EPOCH, ZoneId.systemDefault());
+    private Set<Branch> commitMapBranchSet;
+    private int commitMapCount = 0;
 
     private Map<String, SequencedSet<Hash>> getCommitTitleToCommitsMap(ReadOnlyRepository localRepo, List<Branch> branches) {
-        if (!commitTitleToCommitsInitialized) {
-            try {
+        try {
+            Set<Branch> branchSet = Set.copyOf(branches);
+            if (!branchSet.equals(commitMapBranchSet)) {
+                log.info("Invalidating commitTitleToCommits map for branch set: " + branchSet + " old set: " + commitMapBranchSet);
```

Comment:
Since commitMapBranchSet doesn't have an initial value, this log will also be printed during the initial run. Is this intentional?

Reply:
Yes, kind of. I did note that it would happen and figured it was fine. I don't think it's worth complicating the code further to try to handle the first call differently.

Reply:
I'm also ok with it